firedrake/slate/slate.py

```diff
@@ -0,0 +1,538 @@
+            implemented in the subclass `TensorAdd`.
+        (2) The subtraction of two scalar, vector or matrix expressions. See `TensorSub`.
+        (3) The multiplication of two matrices, the action of a matrix on a vector or
+            the action of a scalar on any other rank tensor. All cases are handled by the
+            `TensorMul` subclass.
+
+    :arg A: a :class:`TensorBase` object. This can be a terminal tensor object
+            (:class:`Tensor`) or any derived expression resulting from any number
+            of linear algebra operations on `Tensor` objects. For example,
+            another instance of a `BinaryOp` object is an acceptable input, or
+            a `UnaryOp` object.
+    :arg B: a :class:`TensorBase` object.
+    """
+
+    def __init__(self, A, B):
+        """Constructor for the BinaryOp class."""
+        super(BinaryOp, self).__init__()
+        self.tensors = A, B
+
+    def coefficients(self):
+        """Returns the expected coefficients of the resulting tensor
+        of performing a binary operation on two tensors. Note that
+        the coefficients are handled the same way for all binary operations.
+        """
+        A, B = self.tensors
+        clist = []
+        A_coeffs = A.coefficients()
+        uniqueAcoeffs = set(A_coeffs)
+        for c in B.coefficients():
+            if c not in uniqueAcoeffs:
+                clist.append(c)
+
+        return tuple(list(A_coeffs) + clist)
+
+    def ufl_domains(self):
+        """Returns the integration domains of the integrals associated with
+        the tensor.
+        """
+        A, B = self.tensors
+        return join_domains(A.ufl_domains() + B.ufl_domains())
+
+    def subdomain_data(self):
+        """Returns a mapping on the tensor:
+        `{domain:{integral_type: subdomain_data}}`.
+        """
+        A, B = self.tensors
+        if A.rank == 0:
+            return B.subdomain_data()
+        elif B.rank == 0:
+            return A.subdomain_data()
+        else:
+            # Join subdomain_data
+            sd_dict = A.subdomain_data()[A.ufl_domain()]
+            for sd_key_B in B.subdomain_data()[B.ufl_domain()].keys():
+                if sd_key_B not in sd_dict.keys():
+                    sd_dict.update({sd_key_B: B.subdomain_data()[B.ufl_domain()][sd_key_B]})
+
+            return {self.ufl_domain(): sd_dict}
+
+    @property
+    def operands(self):
+        """Returns an iterable of the operands of the binary operation."""
+        return self.tensors
+
+    def __str__(self, prec=None):
+        """String representation of the binary operation."""
+
+        ops = {"Addition": '+',
+               "Subtraction": '-',
+               "Multiplication": '*'}
+        if prec is None or self.prec >= prec:
+            par = lambda x: x
+        else:
+            par = lambda x: "(%s)" % x
+        operand1 = self.operands[0].__str__(prec=self.prec)
+        operand2 = self.operands[1].__str__(prec=self.prec)
```

Comment:
So you would rather have all these modified `__str__`'s as their own method and call them in an overloaded `__str__` (without taking the precedence arg `prec`)?